File: src/main/java/cn/xunyard/idea/coding/log/window/AbstractWindowedLogger.java
package cn.xunyard.idea.coding.log.window;

import com.intellij.execution.filters.TextConsoleBuilderFactory;
import com.intellij.execution.ui.ConsoleView;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.wm.ToolWindow;
import com.intellij.openapi.wm.ToolWindowAnchor;
import com.intellij.openapi.wm.ToolWindowManager;
import com.intellij.ui.content.Content;
import com.intellij.ui.content.ContentManager;
import org.jetbrains.annotations.NotNull;

/**
 * @author <a herf="mailto:wuqi@terminus.io">xunyard</a>
 * @date 2019-12-26
 */
public abstract class AbstractWindowedLogger {
    private static final String TOOL_WINDOW_ID = "Coding Helper";

    protected static ConsoleView getConsoleView(@NotNull Project project, @NotNull String displayName) {
        ToolWindow toolWindow = getToolWindow(project);
        Content content = getContent(toolWindow, displayName);

        if (content == null) {
            content = buildContent(project, displayName, toolWindow.getContentManager());
        }

        return (ConsoleView) content.getComponent();
    }

    private static Content getContent(ToolWindow toolWindow, String displayName) {
        Content[] contents = toolWindow.getContentManager().getContents();

        if (contents.length == 0) {
            return null;
        }

        for (Content content : contents) {
            if (content.getDisplayName().equals(displayName)) {
                return content;
            }
        }

        return null;
    }

    private static Content buildContent(Project project, String displayName, ContentManager contentManager) {
        ConsoleView consoleView = TextConsoleBuilderFactory.getInstance().createBuilder(project).getConsole();
        Content content = contentManager.getFactory().createContent(consoleView.getComponent(), displayName, false);
        contentManager.addContent(content);

        return content;
    }

    private static ToolWindow getToolWindow(@NotNull Project project) {
        ToolWindowManager toolWindowManager = ToolWindowManager.getInstance(project);
        ToolWindow toolWindow = toolWindowManager.getToolWindow(TOOL_WINDOW_ID);

        if (toolWindow == null) {
            toolWindow = toolWindowManager.registerToolWindow(TOOL_WINDOW_ID, true, ToolWindowAnchor.BOTTOM);
            toolWindow.activate(null);
        }

        return toolWindow;
    }

    public static void activeToolWindow(@NotNull Project project) {
        ToolWindowManager toolWindowManager = ToolWindowManager.getInstance(project);
        ToolWindow toolWindow = toolWindowManager.getToolWindow(TOOL_WINDOW_ID);

        if (toolWindow == null) {
            toolWindow = toolWindowManager.registerToolWindow(TOOL_WINDOW_ID, true, ToolWindowAnchor.BOTTOM);
        }

        if (!toolWindow.isActive()) {
            toolWindow.activate(null);
        }
    }
}

File: src/main/kotlin/cn/xunyard/idea/coding/log/window/AbstractWindowedLogger.kt
package cn.xunyard.idea.coding.log.window

import com.intellij.execution.filters.TextConsoleBuilderFactory
import com.intellij.execution.ui.ConsoleView
import com.intellij.openapi.project.Project
import com.intellij.openapi.wm.ToolWindow
import com.intellij.openapi.wm.ToolWindowManager
import com.intellij.ui.content.Content
import com.intellij.ui.content.ContentManager

/**
 * @author <a herf="mailto:xunyard@gmail.com">xunyard</a>
 * @date 2019-12-26
 */
abstract class AbstractWindowedLogger {
    companion object {
        /**
         * 同plugin.xml中的id
         */
        private const val TOOL_WINDOW_ID = "Coding Helper"

        @JvmStatic
        internal fun getConsoleView(project: Project, displayName: String): ConsoleView {
            val toolWindow = getToolWindow(project)
            val content = getConsoleView(toolWindow, displayName)
                    ?: buildConsoleView(project, displayName, toolWindow.contentManager)
            return content.component as ConsoleView
        }

        private fun getConsoleView(toolWindow: ToolWindow, displayName: String): Content? {
            val contents = toolWindow.contentManager.contents
            if (contents.isEmpty()) {
                return null
            }
            return contents.firstOrNull { it.displayName == displayName }
        }

        private fun buildConsoleView(project: Project, displayName: String, contentManager: ContentManager): Content {
            val consoleView = TextConsoleBuilderFactory.getInstance().createBuilder(project).console
            return contentManager.factory.createContent(consoleView.component, displayName, false).apply {
                contentManager.addContent(this)
            }
        }

        private fun getToolWindow(project: Project): ToolWindow {
            val toolWindowManager = ToolWindowManager.getInstance(project)
            return toolWindowManager.getToolWindow(TOOL_WINDOW_ID)!!
        }

        fun activeToolWindow(project: Project) {
            val toolWindowManager = ToolWindowManager.getInstance(project)
            val toolWindow = toolWindowManager.getToolWindow(TOOL_WINDOW_ID)!!
            if (!toolWindow.isActive) {
                toolWindow.activate(null)
            }
        }
    }
}